Update intensity methods between instrument and physical units

kakirastern opened the pull request:
Fixes #27. 

Intended to complete the GSoC 2019 IRISpy project before tests are written and perfected for the relevant methods. 

Aim
===
To update intensity conversion methods between instrument and physical units that correct for the time observations were taken, by going through the chain `iris_tools.calculate_photons_per_sec_to_radiance_factor`, `iris_tools.convert_or_undo_photons_per_sec_to_radiance`, and `spectrogaph.IRISSpectrogramCube.convert_to`. 

Essentially, the final stage of the project is to find all the cases in the code base that call `_get_interpolated_effective_area` (or `get_iris_response`) and enable those functions to supply a time and version input `kwarg` that then gets passed onto `_get_interpolated_effective_area` (or `get_iris_response`).

kakirastern commented:
Hi @DanRyanIrish, I have opened this PR to round out our work on the GSoC IRISpy project. And, I intend to open yet another PR later to complete the tests required to launch IRISpy. 

In this PR, I have gotten rid of the `iris_fit = fit_iris_xput` from before, and have replaced this with a higher level `get_iris_response(time_obs, pre_launch, response_version, **kwargs)` instead, as you have suspected I needed to do earlier. 

Please have a look and let me know what needs changing if necessary. I would appreciate it a lot.

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -504,8 +504,7 @@ def _calculate_orbital_wavelength_variation(data_array, date_data_created, slit_
         else:
-            raise TypeError("slit_pixel_range must be tuple of length 2 giving lower and " +
-                            "upper bounds of section of slit over which to average line fits.")
+            raise TypeError("slit_pixel_range must be tuple of length 2 giving lower and " + "upper bounds of section of slit over which to average line fits.")
```
PEP8 compliance:  Lines must be <100 characters.  This line looks like it might be more than that.  If so, break into two lines as previously.

kakirastern commented:
Hi @DanRyanIrish Changes made to the PR according to suggestions. But I may have missed something important. If so, please let me know.

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -435,7 +435,8 @@ def _gaussian1d_on_linear_bg(x, amplitude=None, mean=None, standard_deviation=No
 
 
-def _calculate_orbital_wavelength_variation(data_array, date_data_created, slit_pixel_range=None, spline_smoothing=False, fit_individual_profiles=False, spacecraft_velocity=None, orbital_phase=None, roll_angle=None):
+def _calculate_orbital_wavelength_variation(data_array, date_data_created, \ slit_pixel_range=None, spline_smoothing=False, fit_individual_profiles=False, \
```
the ```\```'s aren't necessary.  If there's a line break but the command is not closed properly, e.g. a bracket is missing or the line ends in a comma, Python automatically assumes it's one line of code stretched over two lines of text.

kakirastern replied:
But I thought this way it would be clearer where the line breaks are... otherwise it could look kind of messy in some places as I have just saw moments ago. Is it okay to keep them?

kakirastern replied:
Okay, I will go ahead and remove the `\`'s then...

DanRyanIrish replied:
It's not in keeping with standard Python coding style unless they are explicitly needed.

kakirastern replied:
Done, all redundant `\`'s removed from `iris_tools.py`

kakirastern replied:
Done the same for `spectrograph.py` as well.

kakirastern commented:
Hi @DanRyanIrish. Anything I will need to add/change before moving on to working on the tests in another PR?

DanRyanIrish commented:
Hi @kakirastern.  Looks like you've now also removed some ```\``` that are in fact necessary.  The rules for line continuation in Python aren't necessarily intuitive but it can be easy to forget that when you're used to them.  So apologies if my instructions were confusing.

I'd recommend reading [this guide](https://developer.rhino3d.com/guides/rhinopython/python-statements/) to learn when ```\``` should be used and when not.

Also it would be a good idea to read the [PEP8 style guide](https://www.python.org/dev/peps/pep-0008/) which is the standard style guide for writing readable Python code.  The section of [maximum line length](https://www.python.org/dev/peps/pep-0008/#maximum-line-length) talks a little about line continuation.

Also note that although the maximum recommended line length is 79, it also states that 99 is acceptable.  We use 99 in IRISpy to allow us to use more lengthy, descriptive variable names.

There may be several places where IRISpy breaks PEP8.  Don't worry about that as that's not the point of this PR and fixing it will make this PR's diff hard to review.  Just use the lessons on the parts of the code you change.

kakirastern commented:
Hi @DanRyanIrish. Sorry, I was rather careless and have not studied the code carefully before removing some of the backward slashes originally there. I have rebased and dropped some of the unnecessary commits.

kakirastern commented:
Hi @DanRyanIrish Do you think I will have enough time to finish the IRISpy code by the deadline? If not, I should focus on that first before returning to this... as this can wait until after.

kakirastern commented:
@DanRyanIrish And I have run into some Travis CI issues... Looks like they are not related to my PR in particular. I might need the tests to be restarted manually later for the CI checks to pass. Would it be possible for you to offer assistance in this regard?

DanRyanIrish commented:
Sure.  Restarted Travis.

DanRyanIrish commented on irispy/spectrograph.py:
```diff
@@ -333,8 +333,7 @@ def convert_to(self, new_unit_type):
             spectral_wcs_index = np.where(np.array(self.wcs.wcs.ctype) == "WAVE")[0][0]
-            spectral_dispersion_per_pixel = self.wcs.wcs.cdelt[spectral_wcs_index] * \
-                                            self.wcs.wcs.cunit[spectral_wcs_index]
+            spectral_dispersion_per_pixel = self.wcs.wcs.cdelt[spectral_wcs_index] * \ self.wcs.wcs.cunit[spectral_wcs_index]
```
The backslash should be removed or ```self.wcs.wcs.cunit[spectral_wcs_index]``` moved to the next line.  Remember that the length of the line must be <100 characters though.

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -801,8 +801,8 @@ def uncalculate_exposure_time_correction(old_data_arrays, old_unit, exposure_tim
 
-def convert_or_undo_photons_per_sec_to_radiance(
-        data_quantities, obs_wavelength, detector_type,
+def convert_or_undo_photons_per_sec_to_radiance(data_quantities,
```
```time_obs``` is in the docstring but missing from the function definition.

DanRyanIrish commented:
It passed!  But weirdly it doesn't register it here.  Perhaps once you address the couple minor comments I just left it will pass.

Have you manually played with these new functions and methods yourself?  If you have and you're happy they work I think we are close to merging!

kakirastern commented:
Hooray, the Travis CI tests passed!

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -802,7 +802,7 @@ def uncalculate_exposure_time_correction(old_data_arrays, old_unit,
 
 def convert_or_undo_photons_per_sec_to_radiance(data_quantities,
-        response_version, obs_wavelength, detector_type,
+        time_obs, obs_wavelength, detector_type,
```
```response_version``` is still required (see [here](https://github.com/sunpy/irispy/pull/117/files/e49ae96b2b60bd5b7a56374a57d7befac1724497..cdd3b20cf1386fc3bffafc033a0e094ae467a271#diff-dee23d5475a6ed5e368938455c92fb2eR862)) and so should be added back in while still keeping ```time_obs```.

DanRyanIrish commented on irispy/spectrograph.py:
```diff
@@ -333,7 +333,8 @@ def convert_to(self, new_unit_type):
             # Get spectral dispersion per pixel.
             spectral_wcs_index = np.where(np.array(self.wcs.wcs.ctype) == "WAVE")[0][0]
-            spectral_dispersion_per_pixel = self.wcs.wcs.cdelt[spectral_wcs_index] * \ self.wcs.wcs.cunit[spectral_wcs_index]
+            spectral_dispersion_per_pixel = self.wcs.wcs.cdelt[spectral_wcs_index] *
```
Removing the ```\``` AND breaking the line is *not* correct.  It is one of the other.  I suggest just adding a ```\``` like this:
```python
spectral_dispersion_per_pixel = self.wcs.wcs.cdelt[spectral_wcs_index] * \
   self.wcs.wcs.cunit[spectral_wcs_index]
```

kakirastern commented:
Sorry @DanRyanIrish about the confusion earlier... Some of the `\`'s were there before I altered them, so I was hesitant to change them at first. But I will be more careful next time. It is still a learning curve for me. I kind of get the feeling that line breaking can be simple, but tricky.

DanRyanIrish commented:
No problem.  It's largely a matter of experience and habit.  Reading through PEP8 and the other article I linked you to will definitely help.

kakirastern commented:
Yeah, the PEP8 Style Guide for Python Code is particularly useful... Thanks for the info @DanRyanIrish!

DanRyanIrish commented:
I think this is looking pretty good.  Is there anything else you are planning to add to his PR?

kakirastern commented:
Nothing to add, but maybe I should check it one more time...

DanRyanIrish commented:
Have you tried the new functionality out yourself?

kakirastern commented:
Caught one bug introduced by me earlier... But other then that it is go for another review.

kakirastern commented:
> Have you tried the new functionality out yourself in a Python session? If not, I'd definitely recommend doing that by reading in a couple spectrograph FITS files using the reader function in spectrograph.py. Then try converting the data to radiance and back to DN using the `convert_to` method. Make sure it doesn't crash.

I want to do that too but I don't think I have the right spectrograph FITS files for the tests. Do you know where I can find some?

kakirastern commented:
@DanRyanIrish Alternatively, I see some existing tests in the code base. Are these sufficient to check for bugs for the new (or time-enabled) methods?

DanRyanIrish commented:
They will probably catch some bugs but not all.  There's nothing like actually playing with it yourself.

DanRyanIrish commented:
@kakirastern Did you try out this functionality yourself?  If so, did you find any bugs?

kakirastern commented:
@DanRyanIrish Yup, am using some FITS files I have to test the code...

kakirastern commented:
Hi @DanRyanIrish, ran into some issues while working on the tests... Am seeing this error coming from `test_iris_tools.py`:

```
==================================== ERRORS ====================================
_______________ ERROR collecting irispy/tests/test_iris_tools.py _______________
irispy/tests/test_iris_tools.py:264: in <module>
    response_version=4)
irispy/iris_tools.py:246: in get_iris_response
    iris_fit_fuv[:, j] = fit_iris_xput(time_obs, iris_response["C_F_TIME"], iris_response["COEFFS_FUV"][j, :, :])
irispy/iris_tools.py:403: in fit_iris_xput
    t_diff = t - t_cal_coeffs
../astropy/astropy/time/core.py:1858: in __rsub__
    out = self.__sub__(other)
../astropy/astropy/time/core.py:1805: in __sub__
    out._set_scale(self.scale)
../astropy/astropy/time/core.py:742: in _set_scale
    jd1, jd2 = conv_func(*args)
../astropy/astropy/_erfa/core.py:15467: in taiutc
    check_errwarn(c_retval, 'taiutc')
../astropy/astropy/_erfa/core.py:89: in check_errwarn
    raise ErfaError('ERFA function "' + func_name + '" yielded ' + emsg)
E   astropy.utils.exceptions.ErfaError: ERFA function "taiutc" yielded 1 of "unacceptable date"

```

kakirastern commented:
The captured output for that is

```
------------------------------- Captured stdout --------------------------------
--------------------------------------------------
Date: Wed Jul 24 11:09:31 2019
User: lahayes
Host: gs671-lhayesmbp
--------------------------------------------------
Format: 12
Architecture: x86_64
Operating System: darwin
IDL Version: 8.6.1
--------------------------------------------------
Successfully read 4 records of which:
 - 1 are of type VARIABLE
 - 1 are of type VERSION
 - 1 are of type TIMESTAMP
--------------------------------------------------
Available variables:
 - iris_response [<class 'numpy.recarray'>]
--------------------------------------------------
--------------------------------------------------
Date: Wed Jul 24 11:10:21 2019
User: lahayes
Host: gs671-lhayesmbp
--------------------------------------------------
Format: 12
Architecture: x86_64
Operating System: darwin
IDL Version: 8.6.1
--------------------------------------------------
Successfully read 4 records of which:
 - 1 are of type VARIABLE
 - 1 are of type VERSION
 - 1 are of type TIMESTAMP
--------------------------------------------------
Available variables:
 - iris_response [<class 'numpy.recarray'>]
--------------------------------------------------
--------------------------------------------------
Date: Wed Jul 24 11:53:46 2019
User: lahayes
Host: gs671-lhayesmbp
--------------------------------------------------
Format: 12
Architecture: x86_64
Operating System: darwin
IDL Version: 8.6.1
--------------------------------------------------
Successfully read 4 records of which:
 - 1 are of type VARIABLE
 - 1 are of type VERSION
 - 1 are of type TIMESTAMP
--------------------------------------------------
Available variables:
 - iris_response [<class 'numpy.recarray'>]
--------------------------------------------------
--------------------------------------------------
Date: Wed Jul 24 12:21:43 2019
User: lahayes
Host: gs671-lhayesmbp
--------------------------------------------------
Format: 12
Architecture: x86_64
Operating System: darwin
IDL Version: 8.6.1
--------------------------------------------------
Successfully read 4 records of which:
 - 1 are of type VARIABLE
 - 1 are of type VERSION
 - 1 are of type TIMESTAMP
--------------------------------------------------
Available variables:
 - iris_response [<class 'numpy.recarray'>]
--------------------------------------------------
```
Any idea where this error is coming from?

kakirastern commented:
On second thought, maybe I should hurry with the `tests` PR and open it soon so we better discuss the details there.

kakirastern commented:
Sorry @DanRyanIrish I hesitated to start on completing the `tests` earlier out of the extra conflict resolution expected... And also just to wait and see if you wanted me to add extra tests. Guess not. So will move on.

DanRyanIrish commented:
> ```
> ==================================== ERRORS ====================================
> _______________ ERROR collecting irispy/tests/test_iris_tools.py _______________
> irispy/tests/test_iris_tools.py:264: in <module>
>     response_version=4)
> irispy/iris_tools.py:246: in get_iris_response
>     iris_fit_fuv[:, j] = fit_iris_xput(time_obs, iris_response["C_F_TIME"], iris_response["COEFFS_FUV"][j, :, :])
> irispy/iris_tools.py:403: in fit_iris_xput
>     t_diff = t - t_cal_coeffs
> ../astropy/astropy/time/core.py:1858: in __rsub__
>     out = self.__sub__(other)
> ../astropy/astropy/time/core.py:1805: in __sub__
>     out._set_scale(self.scale)
> ../astropy/astropy/time/core.py:742: in _set_scale
>     jd1, jd2 = conv_func(*args)
> ../astropy/astropy/_erfa/core.py:15467: in taiutc
>     check_errwarn(c_retval, 'taiutc')
> ../astropy/astropy/_erfa/core.py:89: in check_errwarn
>     raise ErfaError('ERFA function "' + func_name + '" yielded ' + emsg)
> E   astropy.utils.exceptions.ErfaError: ERFA function "taiutc" yielded 1 of "unacceptable date"
> ```

What are the values of ```t``` and ```t_cal_coeffs```?  Looks like something is not in a valid astropy time format.

kakirastern commented:
Yup, it was. And I have fixed it in #118.

kakirastern commented:
So finally I can say yes, I have tested out the code, up to the `convert_to()` part... It has not crashed.

kakirastern commented:
But one problem though when I tried to convert to `"radiance"`:
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-67-8403b243ff48> in <module>
----> 1 new = IRISSpectrogramCube(data=data1, wcs=wcs1, meta=meta0, uncertainty=np.sqrt(np.abs(data1)), unit=iris_tools.DN_UNIT["FUV"], extra_coords=extra_coords0).convert_to("radiance")

~/irispy/irispy/spectrograph.py in convert_to(self, new_unit_type)
    382                 new_data_quantities = iris_tools.convert_or_undo_photons_per_sec_to_radiance(
    383                     (cube.data*cube.unit, cube.uncertainty.array*cube.unit),
--> 384                     obs_wavelength, detector_type, spectral_dispersion_per_pixel, solid_angle)
    385                 new_data = new_data_quantities[0].value
    386                 new_uncertainty = new_data_quantities[1].value

~/irispy/irispy/iris_tools.py in convert_or_undo_photons_per_sec_to_radiance(data_quantities, obs_wavelength, detector_type, spectral_dispersion_per_pixel, solid_angle, undo)
    866                     "of data_quantities. Unit: {2}".format(u.photon/u.s, i, data.unit))
    867     photons_per_sec_to_radiance_factor = calculate_photons_per_sec_to_radiance_factor(
--> 868         obs_wavelength, detector_type, spectral_dispersion_per_pixel, solid_angle)
    869     # Change shape of arrays so they are compatible for broadcasting
    870     # with data and uncertainty arrays.

TypeError: calculate_photons_per_sec_to_radiance_factor() missing 3 required positional arguments: 'detector_type', 'spectral_dispersion_per_pixel', and 'solid_angle'
```

kakirastern commented:
@DanRyanIrish Looks like this code block for convert to "radiance" needs a bit of debugging...

kakirastern commented:
Hi @DanRyanIrish I am staring at the code for `convert_to` right now... What do you suggest would be a good plan of attack for fixing the bug in conversion to `radiance`? Any quick fix possible?

kakirastern commented:
Now after some commits I am seeing the following error instead:

```
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-17-d4db7355bcfa> in <module>
----> 1 new = IRISSpectrogramCube(data=data1, wcs=wcs1, meta=meta0, uncertainty=np.sqrt(np.abs(data1)), unit=iris_tools.DN_UNIT["FUV"], extra_coords=extra_coords0).convert_to("radiance", time_obs=parse_time('2013-09-03', format='utime'))

~/irispy/irispy/spectrograph.py in convert_to(self, new_unit_type, time_obs, response_version)
    387                         (cube.data*cube.unit, cube.uncertainty.array*cube.unit),
    388                          time_obs, response_version, obs_wavelength, detector_type,
--> 389                          spectral_dispersion_per_pixel, solid_angle)
    390                 new_data = new_data_quantities[0].value
    391                 new_uncertainty = new_data_quantities[1].value

~/irispy/irispy/iris_tools.py in convert_or_undo_photons_per_sec_to_radiance(data_quantities, time_obs, response_version, obs_wavelength, detector_type, spectral_dispersion_per_pixel, solid_angle, undo)
    874                     "of data_quantities. Unit: {2}".format(u.photon/u.s, i, data.unit))
    875     photons_per_sec_to_radiance_factor = calculate_photons_per_sec_to_radiance_factor(time_obs,
--> 876         response_version, obs_wavelength, detector_type, spectral_dispersion_per_pixel, solid_angle)
    877     # Change shape of arrays so they are compatible for broadcasting
    878     # with data and uncertainty arrays.

~/irispy/irispy/iris_tools.py in calculate_photons_per_sec_to_radiance_factor(time_obs, reponse_version, wavelength, detector_type, spectral_dispersion_per_pixel, solid_angle)
    928     # Get effective area and interpolate to observed wavelength grid.
    929     eff_area_interp = _get_interpolated_effective_area(time_obs,
--> 930         response_version, detector_type, obs_wavelength=wavelength)
    931     # Return radiometric conversed data assuming input data is in units of photons/s.
    932     return constants.h * constants.c / wavelength / u.photon / spectral_dispersion_per_pixel / eff_area_interp / solid_angle

NameError: name 'response_version' is not defined
```

DanRyanIrish commented:
Looks like it's because of a typo:
```~/irispy/irispy/iris_tools.py in calculate_photons_per_sec_to_radiance_factor(time_obs, reponse_version...```

kakirastern commented:
Thanks @DanRyanIrish Got it working now: 

<img width="1232" alt="Screenshot 2019-08-13 15 22 37" src="https://user-images.githubusercontent.com/26264600/62922553-420c4300-bdde-11e9-929c-47ac162f3a34.png">

kakirastern commented:
Tested locally with versions = 1,2,3,4 and the "raster" FITS file `iris_l2_20130720_043439_4190262172_raster_t000_r00000.fits`. And the `convert_to` utility is now working for conversion into `radiance`.

kakirastern commented:
This PR is ready for a review.

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -139,14 +143,18 @@ def get_iris_response(time_obs=None, pre_launch=False, response_file=None, respo
-            response_version = 4
-    if response_file_set+pre_launch+response_version_set > 1:
+    if pre_launch_set + response_file_set + response_version_set == 2 or \
+        pre_launch_set + response_file_set + response_version_set == 3:
```
Why was this changed? It looks like it does the same job as the previous code except that it has to perform more checks...  Or was there some unexpected bug?

kakirastern replied:
Yeah, I understand. But simply using `pre_launch` would throw an error for some yet to be identified reason.

kakirastern replied:
I was wondering why too, but the new code seems to work just fine.

DanRyanIrish replied:
How odd!  Is it just the use of ```pre_launch```?  For example does this work:
```python
if response_file_set + pre_launch_set + response_version_set > 1:
```
?

kakirastern replied:
Yeah, think it must have been the use of `pre_launch`, but I am not sure why it doesn't work the way it was intended to. It should work supposedly. But doesn't...

DanRyanIrish commented on irispy/iris_tools.py:
```diff
@@ -139,14 +143,18 @@ def get_iris_response(time_obs=None, pre_launch=False, response_file=None, respo
+        try:
+            response_version = 4
+        except Exception:
+            raise("Version number not recognized.")
```
In what case does this exception ever get raised?  Surely ```response_version = 4``` should always work?

kakirastern replied:
Yeah true. I should delete the exception part.

DanRyanIrish commented on irispy/spectrograph.py:
```diff
@@ -308,7 +308,7 @@ def __repr__(self):
+    def convert_to(self, new_unit_type, time_obs=None, response_version=4):
         """
         Converts data, unit and uncertainty attributes to new unit type.
 
```
Don't forget to add an explanation of ```time_obs``` and ```response_version``` to this docstring.

kakirastern commented:
@DanRyanIrish May need to restart Travis CI at some later point in time for the tests to pass

DanRyanIrish commented:
Great job!
